app/ts/contenthash.ts

```diff
@@ -0,0 +1,174 @@
+import { varint } from 'multiformats'
+import { base32 } from 'multiformats/bases/base32'
+import { base36 } from 'multiformats/bases/base36'
+import { base58btc } from 'multiformats/bases/base58'
+import { base64url } from 'multiformats/bases/base64'
+import { CID } from 'multiformats/cid'
+import { create as createDigest, decode as multihashDecode } from 'multiformats/hashes/digest'
+import { assertNever } from './library/utilities.js'
+
+export const codeToName = {
+	0xe3: 'ipfs',
+	0xe4: 'swarm',
+	0xe5: 'ipns',
+	0x01bc: 'onion',
+	0x01bd: 'onion3',
+	0xb19910: 'skynet',
+	0xb29910: 'arweave',
+} as const
+
+export const nameToCode = {
+	ipfs: 0xe3,
+	swarm: 0xe4,
+	ipns: 0xe5,
+	onion: 0x01bc,
+	onion3: 0x01bd,
+	skynet: 0xb19910,
+	arweave: 0xb29910,
+} as const;
+  
+type CodecId = keyof typeof codeToName
+
+const hexStringToBytes = (hex: string): Uint8Array => {
+	let value: string = hex
+	if (value.startsWith('0x')) value = value.slice(2)
+	if (value.length % 2 !== 0) throw new Error('Invalid hex string')
+	const bytes = new Uint8Array(value.length / 2)
+	for (let i = 0; i < value.length; i += 2) {
+		bytes[i / 2] = parseInt(value.slice(i, i + 2), 16)
+	}
+	return bytes
+}
+
+const bytesToHexString = (bytes: Uint8Array): `0x${ string }` => {
+	let hex = ''
+	for (let i = 0; i < bytes.length; i++) {
+		const b = bytes[i]
+		if (b == undefined) throw new Error('length misalign')
+		hex += b.toString(16).padStart(2, '0')
+	}
+	return `0x${ hex }`
+}
+
+const isCryptographicIPNS = (cid: CID): boolean => {
+	try {
+		const { multihash } = cid
+		if (multihash.size < 38) {
+			const mh = multihashDecode(multihash.bytes)
+			if (mh.code === 0x0 && mh.size < 36) return false
+		}
+		return true
+	} catch (_) {
```

Comment:
can be an optional catch binding here if we're not interested with the exception
```suggestion
	} catch {
```